### PR TITLE
Add flow limiter (`step_limiter!`)

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/1900.

#### Main branch

plot

#### This branch

plot

#### Conclusions

Having zero forcing when zero forcing is expected is now enforced properly. In the same way, I'm pretty sure forcing will now never exceed the given flux, and flow should be non-negative where expected.

Having total forcing when the basin is not in the low storage factor regime is harder to enforce, because there is no exact way to infer whether the basin storage passed through the reduction factor regime in the last timestep. A conservative heuristic might suffice here though. 

#### Main branch
- HWS: 9.837521 seconds
- De Dommel: 51.496409 seconds

#### this branch
- HWS: 9.554516 seconds
- De Dommel: 64.802680 seconds

### CODE REVIEW

File: core/src/config.jl [hunk 1]
```diff
@@ -249,6 +249,18 @@ const algorithms = Dict{String, Type}(
     "Euler" => Euler,
 )
 
+"""
+Check whether the given function has a method that accepts the given kwarg.
+Note that it is possible that methods exist that accept :a and :b individually,
+but not both.
+"""
+function function_accepts_kwarg(f, kwarg)::Bool
```
Comment (evetion) [inline core/src/config.jl:257]: Very meta, but nice improvement. Is `kwarg_decl` documented (i.e. public?)

Comment (SouthEndMusic) [reply core/src/config.jl:257]: https://julialang.slack.com/archives/C7T968HRU/p1728982500911379?thread_ts=1728982500.911379&cid=C7T968HRU

Comment (SouthEndMusic) [reply core/src/config.jl:257]: I don't think `Base.kwarg_decl` is documented, I can't find it and it doesn't have a docstring.

**Comment by SouthEndMusic** [comment]
It is indeed a fix for #1900, which is one way in which the flows have to be 'tamed' as discussed in #1897. Even though it is not a global fix, I think this PR is a helpful one, as it enforces properties of the solution which we want to be strictly true for any set of solver settings, e.g.
- Non-negative flows where expected
- Realized user demand should not exceed demand

With the ever increasing norm of the state there is still the possibility that the solution slowly drifts away from the physics in ways we cannot capture in the `step_limiter`. If we want to solve that and keep using the fast solver `QNDF`, we either have to find a way to tame the state size (for instance getting resetting to `0` to work), or document that longer simulations require smaller relative errors.

File: core/src/config.jl (hunk 1 above)
Comment (evetion) [reply core/src/config.jl:257]: Well that's bad, but then again, it seems this method has been there for at least 5 years, so let's use it. Nice find.

**Review by evetion — APPROVED** [review]

**Comment by SouthEndMusic** [comment]
I've added the `Average flow states` commit for the record since I was working on it on this branch, but I'm not going to use it. I feel like it's not worth the added complexity, both conceptually and in the code. My goal now is to fix urgent things with the step limiter, and either stop using the relative error (i.e. hardcoding it to be `0`) or documenting that it should be very small if users note growing inaccuracies for long simulations.

**Comment by SouthEndMusic** [comment]
Closed in favor of https://github.com/Deltares/Ribasim/pull/1911.

**Pull request closed without merging**